### PR TITLE
Fix formatting for multi-line patterns

### LINKED ISSUE #42861: Forced formatting on property patterns makes them unsuitable in nested-expression locations.
For example:

image

This is really an unpleasant experience.  Normally i woudl also be able to change the indentation of the property pattern here.  However, because the formatter forces this, you get warnings that your code is non-compliant.

We should follow the same approach we take for collection/object initializer where we do not mandate where you braces go.  

Note: i'm only asking for the above for *nested* expressions like this.  If you have a prop pattern in something conceptually not nested, then it's fine to force.  for example:

```c#
var x = symbol is INamespaceSymbol
{
};
```

### LINKED ISSUE #46284: Bad formatting for pattern block
Given this code:

```cs
        private static bool IsCallingConventionModifier(CustomModifier modifier)
        {
            var modifierType = ((CSharpCustomModifier)modifier).ModifierSymbol;
            return (object)modifierType.ContainingAssembly == modifierType.ContainingAssembly.CorLibrary
                   && modifierType.Name != "CallConv"
                   && modifierType.Arity == 0
                   && modifierType.Name.StartsWith("CallConv", StringComparison.Ordinal)
                   && modifierType.ContainingNamespace is
                      {
                          Name: "CompilerServices",
                          ContainingNamespace:
                          {
                              Name: "Runtime",
                              ContainingNamespace:
                              {
                                  Name: "System",
                                  ContainingNamespace: { IsGlobalNamespace: true }
                              }
                          }
                      };
        }
```

The formatter wants to do this:

```cs
        private static bool IsCallingConventionModifier(CustomModifier modifier)
        {
            var modifierType = ((CSharpCustomModifier)modifier).ModifierSymbol;
            return (object)modifierType.ContainingAssembly == modifierType.ContainingAssembly.CorLibrary
                   && modifierType.Name != "CallConv"
                   && modifierType.Arity == 0
                   && modifierType.Name.StartsWith("CallConv", StringComparison.Ordinal)
                   && modifierType.ContainingNamespace is
            {
                Name: "CompilerServices",
                ContainingNamespace:
                {
                    Name: "Runtime",
                    ContainingNamespace:
                    {
                        Name: "System",
                        ContainingNamespace: { IsGlobalNamespace: true }
                    }
                }
            };
        }
```

This is not very readable, as the braces are not in line with the actual thing being pattern-matched against.

### PR DESCRIPTION
**Pull request by sharwell**
Fixes #42861
Fixes #46284

### CODE REVIEW

File: src/Workspaces/SharedUtilitiesAndExtensions/Compiler/Core/Formatting/BottomUpBaseIndentationFinder.cs [hunk 1]
```diff
@@ -167,7 +171,20 @@ internal class BottomUpBaseIndentationFinder
                 }
 
                 var baseIndentation = tokenColumnGetter(baseToken);
-                return Math.Max(0, baseIndentation + (indentationLevel + operation.IndentationDeltaOrPosition) * _indentationSize);
+                var delta = operation.IndentationDeltaOrPosition;
+                if (operation.Option.IsOn(IndentBlockOption.IndentIfConditionOfAnchorToken))
+                {
+                    if (_syntaxFacts.IsOnIfStatementHeader(root, position, out var conditionStatement)
+                        || _syntaxFacts.IsOnWhileStatementHeader(root, position, out conditionStatement))
```
Comment (CyrusNajmabadi) [inline src/Workspaces/SharedUtilitiesAndExtensions/Compiler/Core/Formatting/BottomUpBaseIndentationFinder.cs:178]: might be worth doc'ing why these two .

File: src/Workspaces/SharedUtilitiesAndExtensions/Compiler/Core/Formatting/Rules/Operations/IndentBlockOption.cs [hunk 2]
```diff
@@ -37,6 +37,19 @@ internal enum IndentBlockOption
         /// <summary>
         /// Mask for position options
         /// </summary>
-        PositionMask = RelativeToFirstTokenOnBaseTokenLine | RelativePosition | AbsolutePosition
+        PositionMask = RelativeToFirstTokenOnBaseTokenLine | RelativePosition | AbsolutePosition,
```
Comment (CyrusNajmabadi) [inline src/Workspaces/SharedUtilitiesAndExtensions/Compiler/Core/Formatting/Rules/Operations/IndentBlockOption.cs:40]: i think this should actually doc why this doesn't cover the item below.

Comment (sharwell) [reply src/Workspaces/SharedUtilitiesAndExtensions/Compiler/Core/Formatting/Rules/Operations/IndentBlockOption.cs:44]: Indirectly documented. Let me know if it works.

Comment (CyrusNajmabadi) [reply src/Workspaces/SharedUtilitiesAndExtensions/Compiler/Core/Formatting/Rules/Operations/IndentBlockOption.cs:44]: sorry, not seeing it, even after rereviewing things.

Comment (sharwell) [reply src/Workspaces/SharedUtilitiesAndExtensions/Compiler/Core/Formatting/Rules/Operations/IndentBlockOption.cs:44]: `IndentIfConditionOfAnchorToken` is not a primary behavior.

Comment (CyrusNajmabadi) [reply src/Workspaces/SharedUtilitiesAndExtensions/Compiler/Core/Formatting/Rules/Operations/IndentBlockOption.cs:44]: this should be documented.   the problem with the code/comemtns as they exist today is that i cannot come back and determine if this is an intentional or accidental omission.  i.e. the comment says "specifies one of the position options to indicate the primary behavior for the operation."  First, i don't know what "primary behavior" means.  Second, i have no way to tell if 'IndentIfConditionOfAnchorToken' is not primary, and thus is appropraite not in this list, or if it is primary, and this is just a mistake.

This is a prime example of why more is more when it comes to certain types of comments.  Just being explicit about intentions here goes a long way for future maintainability.

**Pull request merged**